---
title: "Combined ONT Transcript-Level Spatial Analysis"
author: "Dr Mustafa Elshani"
date: "`r Sys.Date()`"
output: html_document
---

# Combined DMSO vs NUC-7738 Transcript-Level Spatial Transcriptomics

Integration and differential transcript usage analysis of control (DMSO) and NUC-7738 treated samples.

---

# 1. Setup

```{r setup, message=FALSE, warning=FALSE}
library(Giotto)
library(dplyr)
library(ggplot2)
library(patchwork)
library(cowplot)
library(RColorBrewer)
library(terra)
library(exifr)
library(data.table)
library(harmony)
library(scales)
```

# 2. Configuration

```{r config}
# Python path
python_path <- "/home/mustafa/miniforge3/envs/giotto_env/bin/python"

# Results folder
results_folder <- "./results/SpatialTrans_combined_ONTtrans/ONTtrans"
supplementary_figures <- "./figures/paper_figures/supplementary"

# Create directories
dir.create(results_folder, recursive = TRUE, showWarnings = FALSE)
dir.create(supplementary_figures, recursive = TRUE, showWarnings = FALSE)

# Giotto instructions
instrs <- createGiottoInstructions(
  save_plot = TRUE, 
  save_dir = results_folder, 
  show_plot = FALSE, 
  python_path = python_path
)

# Save parameters
base_save_parameters <- list(
  save_dir = results_folder,
  save_format = "png",
  units = "in",
  base_width = 10,
  base_height = 10,
  dpi = 600
)
```

# 3. Load Giotto Objects

```{r load-objects}
SpatialTrans_DMSO_ONTtrans <- loadGiotto("./GiottoSuite_DMSO_ONTtrans")
SpatialTrans_30uM7738_ONTtrans <- loadGiotto("./GiottoSuite_30uM7738_ONTtrans")
```

# 4. QC: Verify Loaded Objects

```{r qc-dmso}
spatPlot2D(gobject = SpatialTrans_DMSO_ONTtrans, 
           cell_color = "in_tissue",
           show_image = TRUE, 
           point_size = 2.5,
           cell_color_code = c("0" = "lightgrey", "1" = "blue"),
           point_alpha = 0.5)
```

```{r qc-nuc7738}
spatPlot2D(gobject = SpatialTrans_30uM7738_ONTtrans, 
           cell_color = "in_tissue",
           show_image = TRUE, 
           point_size = 2.5,
           cell_color_code = c("0" = "lightgrey", "1" = "blue"),
           point_alpha = 0.5)
```

# 5. Join Giotto Objects

```{r join-objects}
SpatialTrans_combined_ONTtrans <- joinGiottoObjects(
  gobject_list = list(SpatialTrans_DMSO_ONTtrans, SpatialTrans_30uM7738_ONTtrans),
  gobject_names = c("DMSO", "30uM7738"),
  join_method = "shift", 
  x_padding = -10000
)

print(head(pDataDT(SpatialTrans_combined_ONTtrans)))
```

```{r plot-combined}
spatPlot2D(gobject = SpatialTrans_combined_ONTtrans, 
           cell_color = "in_tissue", 
           cell_color_code = c("0" = "lightgrey", "1" = "blue"),
           show_image = TRUE, 
           image_name = c("DMSO-image", "30uM7738-image"), 
           point_size = 1, 
           point_alpha = 0.5,
           save_param = base_save_parameters)
```

# 6. Preprocessing

```{r preprocess}
# Subset in-tissue spots
metadata <- pDataDT(SpatialTrans_combined_ONTtrans)
in_tissue_barcodes <- metadata[in_tissue == 1]$cell_ID
SpatialTrans_combined_ONTtrans <- subsetGiotto(SpatialTrans_combined_ONTtrans, 
                                                cell_ids = in_tissue_barcodes)

# Filter
SpatialTrans_combined_ONTtrans <- filterGiotto(
  gobject = SpatialTrans_combined_ONTtrans,
  expression_threshold = 1,
  feat_det_in_min_cells = 4,
  min_det_feats_per_cell = 100,
  expression_values = "raw",
  verbose = TRUE
)

# Normalize
SpatialTrans_combined_ONTtrans <- normalizeGiotto(
  gobject = SpatialTrans_combined_ONTtrans, 
  scalefactor = 6000
)

# Add statistics
SpatialTrans_combined_ONTtrans <- addStatistics(
  gobject = SpatialTrans_combined_ONTtrans, 
  expression_values = "raw"
)
```

```{r plot-filtered}
spatPlot2D(gobject = SpatialTrans_combined_ONTtrans, 
           cell_color = "nr_feats", 
           color_as_factor = FALSE, 
           point_size = 3, 
           show_image = TRUE, 
           image_name = c("DMSO-image", "30uM7738-image"),
           save_param = base_save_parameters)
```

# 7. Clustering (Pre-Harmony)

```{r clustering-pca}
# HVF and PCA
SpatialTrans_combined_ONTtrans <- calculateHVF(gobject = SpatialTrans_combined_ONTtrans)
SpatialTrans_combined_ONTtrans <- runPCA(gobject = SpatialTrans_combined_ONTtrans, 
                                          center = TRUE, 
                                          scale_unit = TRUE)

# sNN network
SpatialTrans_combined_ONTtrans <- createNearestNetwork(
  gobject = SpatialTrans_combined_ONTtrans,
  dim_reduction_to_use = "pca", 
  dim_reduction_name = "pca",
  dimensions_to_use = 1:10, 
  k = 15
)

# Leiden clustering
SpatialTrans_combined_ONTtrans <- doLeidenCluster(
  gobject = SpatialTrans_combined_ONTtrans, 
  resolution = 0.2,
  n_iterations = 200
)

# UMAP
SpatialTrans_combined_ONTtrans <- runUMAP(SpatialTrans_combined_ONTtrans)
```

```{r plot-pre-harmony}
spatDimPlot2D(gobject = SpatialTrans_combined_ONTtrans,
              cell_color = "leiden_clus", 
              show_image = TRUE, 
              image_name = c("DMSO-image", "30uM7738-image"),
              save_param = base_save_parameters)
```

# 8. Harmony Batch Correction

```{r harmony}
SpatialTrans_combined_ONTtrans <- runGiottoHarmony(
  SpatialTrans_combined_ONTtrans, 
  vars_use = "list_ID",
  do_pca = FALSE,
  sigma = 0.1,
  theta = 2,
  lambda = 1,
  nclust = NULL
)
```

# 9. Clustering (Post-Harmony)

```{r clustering-harmony}
# sNN network on harmony
SpatialTrans_combined_ONTtrans <- createNearestNetwork(
  gobject = SpatialTrans_combined_ONTtrans,
  dim_reduction_to_use = "harmony",
  dim_reduction_name = "harmony", 
  name = "NN.harmony",
  dimensions_to_use = 1:10, 
  k = 15
)

# Leiden clustering
SpatialTrans_combined_ONTtrans <- doLeidenCluster(
  gobject = SpatialTrans_combined_ONTtrans,
  network_name = "NN.harmony", 
  resolution = 0.2, 
  n_iterations = 1000, 
  name = "leiden_harmony"
)

# UMAP
SpatialTrans_combined_ONTtrans <- runUMAP(
  SpatialTrans_combined_ONTtrans, 
  dim_reduction_name = "harmony", 
  dim_reduction_to_use = "harmony", 
  name = "umap_harmony"
)
```

```{r plot-harmony-clusters}
spatDimPlot2D(gobject = SpatialTrans_combined_ONTtrans,
              dim_reduction_to_use = "umap", 
              dim_reduction_name = "umap_harmony",
              cell_color = "leiden_harmony", 
              image_name = c("DMSO-image", "30uM7738-image"),
              spat_point_size = 1, 
              save_param = base_save_parameters)
```

```{r plot-tissue-contribution}
spatDimPlot2D(gobject = SpatialTrans_combined_ONTtrans,
              dim_reduction_to_use = "umap", 
              dim_reduction_name = "umap_harmony",
              cell_color = "list_ID", 
              save_param = base_save_parameters)
```

# 10. Pseudobulk Differential Expression (scran)

```{r scran-de}
scran_results <- findMarkers_one_vs_all(
  SpatialTrans_combined_ONTtrans,
  spat_unit = "cell",
  feat_type = "rna",
  method = "scran",
  expression_values = "normalized",
  cluster_column = "list_ID",
  min_feats = 2
)

top_genes <- scran_results[, head(.SD, 20), by = "cluster"]$feats

# Save results
fwrite(scran_results, 
       file = file.path(supplementary_figures, "normilized_PseudoBulk_DE_DMSOvsNUC7738_trans_scran.csv"))
```

```{r heatmap-de}
plotMetaDataHeatmap(SpatialTrans_combined_ONTtrans, 
                    selected_feats = top_genes, 
                    metadata_cols = "list_ID",
                    save_param = modifyList(base_save_parameters, list(base_width = 5, base_height = 20)))
```

# 11. Spatial Feature Plots

```{r spatial-clusters-by-sample}
spatPlot2D(SpatialTrans_combined_ONTtrans,
           cell_color = "leiden_clus",
           group_by = "list_ID",
           point_size = 2,
           show_legend = FALSE,
           save_param = modifyList(base_save_parameters, list(base_width = 10, base_height = 8)))
```

```{r spatial-transcripts}
custom_feats <- c("ENST00000646664", "ENST00000009530", "ENST00000398752", 
                  "ENST00000394936", "ENST00000636580", "ENST00000361453")

DMSOvsNUC7738_genes <- spatFeatPlot2D(
  SpatialTrans_combined_ONTtrans,
  expression_values = "scaled",
  feats = custom_feats,
  point_size = 1,
  point_alpha = 0.7,
  show_image = TRUE,
  cow_n_col = 2,
  gradient_midpoint = 0,
  image_name = c("DMSO-image", "30uM7738-image"),
  save_param = modifyList(base_save_parameters, list(base_width = 10, base_height = 8))
)

ggsave(
  filename = file.path(supplementary_figures, "figure_4_wholesection_trans.png"),
  plot = DMSOvsNUC7738_genes,
  width = 10, height = 8, dpi = 600
)
```

# 12. Save Combined Object

```{r save-object}
saveGiotto(SpatialTrans_combined_ONTtrans, "./GiottoSuite_combined_ONTtrans", overwrite = TRUE)
```

# 13. DRIMSeq Differential Transcript Usage

## 13.1 Prepare Sample Metadata

```{r drimseq-samples}
cell_metadata <- pDataDT(SpatialTrans_combined_ONTtrans)
samples_df <- data.frame(
  sample_id = cell_metadata$cell_ID,
  condition = cell_metadata$list_ID,
  stringsAsFactors = FALSE
)
samples_df$condition <- factor(samples_df$condition)
```

## 13.2 Extract Expression Matrix

```{r drimseq-expression}
cts_giotto <- getExpression(SpatialTrans_combined_ONTtrans, values = "normalized")
cts_sparse <- cts_giotto@exprMat
cts_mat <- as.matrix(cts_sparse)
cts_df <- as.data.frame(cts_mat)
```

## 13.3 Gene-Transcript Mapping

```{r drimseq-txdb}
library(GenomicFeatures)
library(rtracklayer)

gtf_path <- "./References/gencode.v47.primary_assembly.annotation.gtf.gz"
txdb <- makeTxDbFromGFF(gtf_path)
txdf <- select(txdb, keys(txdb, "GENEID"), "TXNAME", "GENEID")
txdf$GENEID <- sub("\\..*", "", txdf$GENEID)
txdf$TXNAME <- sub("\\..*", "", txdf$TXNAME)

# Match transcripts
txdf <- txdf[txdf$TXNAME %in% rownames(cts_df), ]
cts_df <- cts_df[rownames(cts_df) %in% txdf$TXNAME, ]
txdf <- txdf[match(rownames(cts_df), txdf$TXNAME), ]
```

## 13.4 Run DRIMSeq

```{r drimseq-run}
library(DRIMSeq)

counts <- data.frame(
  gene_id = txdf$GENEID,
  feature_id = txdf$TXNAME,
  cts_df,
  check.names = FALSE
)

d <- dmDSdata(counts = counts, samples = samples_df)
design_full <- model.matrix(~condition, data = DRIMSeq::samples(d))

d <- dmPrecision(d, design = design_full)
d <- dmFit(d, design = design_full)
d <- dmTest(d, coef = "conditionDMSO")

res <- DRIMSeq::results(d)
res.txp <- DRIMSeq::results(d, level = "feature")
```

## 13.5 Annotate with Gene Names

```{r drimseq-annotate}
library(biomaRt)

mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
genes <- unique(res.txp$gene_id)

gene_mapping <- getBM(
  attributes = c("ensembl_gene_id", "external_gene_name"),
  filters = "ensembl_gene_id",
  values = genes,
  mart = mart
)

res.txp <- merge(res.txp, gene_mapping,
                 by.x = "gene_id",
                 by.y = "ensembl_gene_id",
                 all.x = TRUE)
res.txp <- res.txp[order(res.txp$pvalue), ]

# Save results
fwrite(res.txp, file = file.path(supplementary_figures, "Supplementary Table S4.csv"))
```

# 14. Helper Functions for Isoform Visualization

```{r helper-spatinsitu}
spatinsitu_feat_workaround <- function(
    gobject, feat, 
    expression_values = "scaled", 
    spat_unit = NULL, 
    feat_type = NULL, 
    ...) {
  expr_table <- spatValues(gobject,
                           spat_unit = spat_unit, 
                           feat_type = feat_type,
                           expression_values = expression_values,
                           feats = feat)
  gobject <- addCellMetadata(gobject,
                             spat_unit = spat_unit,
                             feat_type = feat_type,
                             new_metadata = expr_table,
                             by_column = TRUE)
  p <- list(gobject = gobject, polygon_fill = feat, ...)
  do.call(spatInSituPlotPoints, p)
}
```

```{r helper-spatial-plot}
create_spatial_plot <- function(giotto_obj, gene, image_name, xlim, ylim, 
                                 x_right, y_bottom, show_title = TRUE) {
  scale_bar_length_microns <- 150
  scale_bar_length_pixels <- scale_bar_length_microns / microns_per_pixel
  
  p <- spatinsitu_feat_workaround(
    giotto_obj, 
    feat = gene,
    show_image = TRUE, 
    image_name = image_name,
    return_plot = TRUE,
    save_plot = FALSE,
    polygon_alpha = 0.7,
    polygon_line_size = 0.1,
    polygon_color = "#333030",
    background_color = "white", 
    xlim = xlim, 
    ylim = ylim
  ) +
    ggplot2::theme_classic() +
    ggplot2::theme(
      axis.line = ggplot2::element_line(color = "black"),
      axis.title.x = ggplot2::element_text(size = 19, face = "bold"),
      axis.title.y = ggplot2::element_text(size = 19, face = "bold"),
      axis.text.x = ggplot2::element_text(size = 18),
      axis.text.y = ggplot2::element_text(size = 14, angle = 90, hjust = 0.5),
      plot.title = element_text(hjust = 0.5, size = 24, face = "bold"),
      legend.title = element_blank(),
      legend.text = element_text(size = 16),
      legend.key.size = unit(1.0, "cm"),
      plot.margin = margin(0, 2, 0, 2)
    ) +
    ggplot2::labs(
      x = "x coordinates",
      y = "y coordinates",
      title = if(show_title) gene else NULL
    ) +
    ggplot2::scale_fill_gradient(low = "white", high = "red", limits = c(0, 2), oob = scales::squish) +
    ggplot2::annotate("segment",
      x = x_right, xend = x_right + scale_bar_length_pixels,
      y = y_bottom + 400, yend = y_bottom + 400,
      color = "black", linewidth = 2, lineend = "round"
    ) +
    ggplot2::annotate("text",
      x = x_right + scale_bar_length_pixels / 2, y = y_bottom + 200,
      label = paste0(scale_bar_length_microns, " µm"),
      color = "black", size = 5, hjust = 0.5, fontface = "bold"
    )
  
  return(p)
}
```

```{r helper-isoform-plot}
library(ggtranscript)

generate_isoform_plot <- function(genes, txdb) {
  isoform_plots <- list()
  transcript_exons <- exonsBy(txdb, by = "tx", use.names = TRUE)
  
  for (gene in genes) {
    specific_exons <- transcript_exons[[which(names(transcript_exons) == gene)]]
    exon_df <- as.data.frame(specific_exons) %>%
      dplyr::mutate(transcript_name = "", type = "exon")

    isoform_plot <- ggplot(exon_df, aes(xstart = start, xend = end, y = transcript_name)) +
      geom_range(fill = "steelblue", color = "black", alpha = 0.7) +
      geom_intron(
        data = to_intron(exon_df, "transcript_name"),
        aes(strand = strand),
        arrow = grid::arrow(angle = 30, length = unit(0.1, "inches"))
      ) +
      theme_minimal() +
      theme(
        axis.title.y = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()
      ) +
      labs(x = "Genomic Coordinates")
    
    isoform_plots[[gene]] <- isoform_plot
  }
  
  plot_grid(plotlist = isoform_plots, ncol = length(genes))
}
```

```{r helper-row-labels}
row_labels <- function(label_text) {
  ggplot() +
    ggplot2::annotate("text", x = 0.5, y = 0.5, label = label_text, 
                      size = 12, fontface = "bold", angle = 90) +
    theme_void()
}
```

# 15. Load TxDb for Isoform Visualization

```{r load-txdb-ensembl, message=FALSE, warning=FALSE}
library(GenomicFeatures)
options(ucscChromosomeNames = FALSE)
txdb_ensembl <- makeTxDbFromEnsembl("Homo sapiens", release = 111)
```

# 16. Isoform Spatial Visualization (Example: UQCRQ)

```{r isoform-visualization, fig.width=18, fig.height=16}
# Function to get transcripts
get_transcripts <- function(gene_name, res.txp) {
  res.txp %>%
    filter(external_gene_name == gene_name, adj_pvalue < 0.05) %>% 
    pull(feature_id)
}

# Gene of interest
gene_name <- "UQCRQ"
genes <- get_transcripts(gene_name, res.txp)

# Calculate microns per pixel
tiff_file_path <- "./DMSO/spaceranger_out_run2/outs/spatial/tissue_fullres_image.tif"
metadata_tiff <- read_exif(tiff_file_path)
microns_per_pixel <- 10000 / metadata_tiff$XResolution

scale_bar_length_microns <- 150
scale_bar_length_pixels <- scale_bar_length_microns / microns_per_pixel

# Function to get significance stars
get_stars <- function(p) {
  if (is.na(p)) return("")
  if (p < 1e-5) return("***")
  else if (p < 1e-3) return("**")
  else if (p < 0.05) return("*")
  else return("ns")
}

# Extract expression data for violin plots
expr_matrix <- getExpression(SpatialTrans_combined_ONTtrans, values = "scaled", output = "matrix")
cell_meta <- pDataDT(SpatialTrans_combined_ONTtrans)
cell_meta$condition <- ifelse(grepl("DMSO|dmso", cell_meta$list_ID, ignore.case = TRUE), "DMSO", "NUC-7738")

# Create violin data
violin_data <- data.frame()
for (gene in genes) {
  if (gene %in% rownames(expr_matrix)) {
    gene_expr <- as.numeric(expr_matrix[gene, ])
    temp_df <- data.frame(
      gene = gene,
      expression = gene_expr,
      cell_id = colnames(expr_matrix),
      condition = cell_meta$condition
    )
    violin_data <- rbind(violin_data, temp_df)
  }
}
violin_data$gene <- factor(violin_data$gene, levels = genes)
violin_data$condition <- factor(violin_data$condition, levels = c("DMSO", "NUC-7738"))

# Get isoform structures
transcript_exons <- exonsBy(txdb_ensembl, by = "tx", use.names = TRUE)

# Create all individual plots first
dmso_plots <- list()
nuc_plots <- list()
isoform_plots <- list()
violin_plots <- list()

for (i in seq_along(genes)) {
  gene <- genes[i]
  
  # Control (DMSO) spatial plot
  dmso_plots[[gene]] <- create_spatial_plot(
    SpatialTrans_combined_ONTtrans, gene, "DMSO-image", 
    xlim = c(13600, 22000), ylim = c(-17200, -13800), 
    x_right = 22000 - scale_bar_length_pixels - 20, 
    y_bottom = -17200 + 80, show_title = TRUE
  ) + theme(plot.margin = margin(2, 2, 2, 2))
  
  # NUC-7738 spatial plot
  nuc_plots[[gene]] <- create_spatial_plot(
    SpatialTrans_combined_ONTtrans, gene, "30uM7738-image", 
    xlim = c(40900, 49900), ylim = c(-25200, -21500), 
    x_right = 49900 - scale_bar_length_pixels - 20, 
    y_bottom = -25200 + 80, show_title = FALSE
  ) + theme(plot.margin = margin(2, 2, 2, 2))
  
  # Isoform structure
  specific_exons <- transcript_exons[[which(names(transcript_exons) == gene)]]
  exon_df <- as.data.frame(specific_exons) %>%
    dplyr::mutate(transcript_name = "", type = "exon")
  
  isoform_plots[[gene]] <- ggplot(exon_df, aes(xstart = start, xend = end, y = transcript_name)) +
    geom_range(fill = "steelblue", color = "black", alpha = 0.7) +
    geom_intron(
      data = to_intron(exon_df, "transcript_name"),
      aes(strand = strand),
      arrow = grid::arrow(angle = 30, length = unit(0.1, "inches"))
    ) +
    theme_minimal() +
    theme(
      axis.title.y = element_blank(),
      axis.text.y = element_blank(),
      axis.ticks.y = element_blank(),
      axis.title.x = element_text(size = 14, face = "bold"),
      axis.text.x = element_text(size = 12),
      plot.margin = margin(2, 2, 2, 2)
    ) +
    labs(x = "Genomic Coordinates")
  
  # Violin plot
  gene_data <- violin_data[violin_data$gene == gene, ]
  max_expr <- max(gene_data$expression, na.rm = TRUE)
  
  p_val <- res.txp[res.txp$feature_id == gene, "adj_pvalue"]
  if (length(p_val) > 0) p_val <- p_val[1] else p_val <- NA
  stars <- get_stars(p_val)
  
  violin_plots[[gene]] <- ggplot(gene_data, aes(x = condition, y = expression, fill = condition)) +
    geom_violin(trim = FALSE, alpha = 0.7, scale = "width") +
    geom_jitter(width = 0.15, size = 0.3, alpha = 0.3) +
    geom_boxplot(width = 0.15, fill = "white", outlier.shape = NA, alpha = 0.8) +
    scale_fill_manual(values = c("DMSO" = "#377EB8", "NUC-7738" = "#FF7F00")) +
    annotate("segment", x = 1, xend = 2, y = max_expr * 1.20, yend = max_expr * 1.20, linewidth = 0.5) +
    annotate("segment", x = 1, xend = 1, y = max_expr * 1.15, yend = max_expr * 1.20, linewidth = 0.5) +
    annotate("segment", x = 2, xend = 2, y = max_expr * 1.15, yend = max_expr * 1.20, linewidth = 0.5) +
    annotate("text", x = 1.5, y = max_expr * 1.26, label = stars, size = 5, fontface = "bold") +
    scale_y_continuous(expand = expansion(mult = c(0.05, 0.32))) +
    theme_classic() +
    theme(
      axis.line = element_line(color = "black", linewidth = 0.8),
      axis.title.x = element_blank(),
      axis.title.y = element_text(size = 16, face = "bold"),
      axis.text.x = element_text(size = 18, face = "bold"),
      axis.text.y = element_text(size = 12),
      legend.position = "none",
      plot.margin = margin(2, 2, 2, 2)
    ) +
    labs(y = "Norm. Expr.")
}

# Create row labels (vertical text)
label_control <- ggplot() +
  annotate("text", x = 0.5, y = 0.5, label = "Control (DMSO)", 
           size = 8, fontface = "bold", angle = 90) +
  theme_void() + theme(plot.margin = margin(0, 0, 0, 0))

label_nuc <- ggplot() +
  annotate("text", x = 0.5, y = 0.5, label = "NUC-7738", 
           size = 8, fontface = "bold", angle = 90) +
  theme_void() + theme(plot.margin = margin(0, 0, 0, 0))

label_A <- ggdraw() + draw_label("A", x = 0.5, y = 0.9, size = 28, fontface = "bold")
label_B <- ggdraw() + draw_label("B", x = 0.5, y = 0.9, size = 28, fontface = "bold")
label_C <- ggdraw() + draw_label("C", x = 0.5, y = 0.9, size = 28, fontface = "bold")

# Row 1: A + Control (DMSO) label + DMSO spatial plots
dmso_row_plots <- plot_grid(plotlist = dmso_plots, ncol = length(genes))
row1 <- plot_grid(
  plot_grid(label_A, label_control, ncol = 1, rel_heights = c(0.15, 1)),
  dmso_row_plots,
  ncol = 2, rel_widths = c(0.06, 1)
)

# Row 2: NUC-7738 label + NUC spatial plots
nuc_row_plots <- plot_grid(plotlist = nuc_plots, ncol = length(genes))
row2 <- plot_grid(
  label_nuc,
  nuc_row_plots,
  ncol = 2, rel_widths = c(0.06, 1)
)

# Combine spatial rows (Control + NUC) with minimal spacing
spatial_combined <- plot_grid(
  row1, row2,
  ncol = 1, rel_heights = c(1, 1)
)

# Row 3: B + isoform structures
isoform_row_plots <- plot_grid(plotlist = isoform_plots, ncol = length(genes))
row3 <- plot_grid(
  label_B,
  isoform_row_plots,
  ncol = 2, rel_widths = c(0.06, 1)
)

# Row 4: C + violin plots
violin_row_plots <- plot_grid(plotlist = violin_plots, ncol = length(genes))
row4 <- plot_grid(
  label_C,
  violin_row_plots,
  ncol = 2, rel_widths = c(0.06, 1)
)

# Combine all rows
main_content <- plot_grid(
  spatial_combined, row3, row4,
  ncol = 1, rel_heights = c(2, 0.3, 0.9)
)

# Add title at top
title_plot <- ggdraw() + 
  draw_label(paste0(gene_name, " Isoform Expression"), 
             x = 0.5, y = 0.5, size = 22, fontface = "bold")

complete_plot <- plot_grid(
  title_plot, main_content,
  ncol = 1, rel_heights = c(0.03, 1)
)

# Save as PNG (main figures folder)
ggsave(
  filename = file.path("./figures/paper_figures", paste0("figure_5_combined_DMSO_NUC7738_", gene_name, "_isoform.png")),
  plot = complete_plot,
  width = 16 , height = 11, dpi = 600, units = "in"
)

# Save as PDF (main figures folder)
ggsave(
  filename = file.path("./figures/paper_figures", paste0("figure_5_combined_DMSO_NUC7738_", gene_name, "_isoform.pdf")),
  plot = complete_plot,
  width =16, height = 11, device = cairo_pdf
)

print(complete_plot)
```

# 17. Supplementary: Whole-Section Spatial Plots (UQCRQ Isoforms)

```{r whole-section-uqcrq, fig.width=22, fig.height=10}
# Create whole-section plots for each UQCRQ transcript (non-cropped)
whole_section_plots <- list()

# Calculate scale bar for whole section
scale_bar_whole <- 150 / microns_per_pixel

for (i in seq_along(genes)) {
  gene <- genes[i]
  
  # DMSO whole section
  dmso_whole <- spatinsitu_feat_workaround(
    SpatialTrans_combined_ONTtrans, 
    feat = gene,
    show_image = TRUE, 
    image_name = "DMSO-image",
    return_plot = TRUE,
    save_plot = FALSE,
    polygon_alpha = 0.7,
    polygon_line_size = 0.1,
    polygon_color = "#333030",
    background_color = "white"
  ) +
    ggplot2::theme_classic() +
    ggplot2::theme(
      axis.line = ggplot2::element_line(color = "black"),
      axis.title.x = ggplot2::element_text(size = 24, face = "bold"),
      axis.title.y = ggplot2::element_text(size = 24, face = "bold"),
      axis.text.x = ggplot2::element_text(size = 18),
      axis.text.y = ggplot2::element_text(size = 18),
      plot.title = element_text(hjust = 0.5, size = 22, face = "bold"),
      legend.title = element_blank(),
      legend.text = element_text(size = 18),
      legend.key.size = unit(1.2, "cm")
    ) +
    ggplot2::labs(
      x = "x coordinates",
      y = "y coordinates",
      title = paste0(gene, " - Control (DMSO)")
    ) +
    ggplot2::scale_fill_gradient(low = "white", high = "red", limits = c(0, 2), oob = scales::squish) +
    ggplot2::annotate("segment",
      x = 5000, xend = 5000 + scale_bar_whole,
      y = -28000, yend = -28000,
      color = "black", linewidth = 2, lineend = "round"
    ) +
    ggplot2::annotate("text",
      x = 5000 + scale_bar_whole / 2, y = -28500,
      label = "150 µm",
      color = "black", size = 8, hjust = 0.5, fontface = "bold"
    )
  
  # NUC-7738 whole section
  nuc_whole <- spatinsitu_feat_workaround(
    SpatialTrans_combined_ONTtrans, 
    feat = gene,
    show_image = TRUE, 
    image_name = "30uM7738-image",
    return_plot = TRUE,
    save_plot = FALSE,
    polygon_alpha = 0.7,
    polygon_line_size = 0.1,
    polygon_color = "#333030",
    background_color = "white"
  ) +
    ggplot2::theme_classic() +
    ggplot2::theme(
      axis.line = ggplot2::element_line(color = "black"),
      axis.title.x = ggplot2::element_text(size = 24, face = "bold"),
      axis.title.y = ggplot2::element_text(size = 24, face = "bold"),
      axis.text.x = ggplot2::element_text(size = 18),
      axis.text.y = ggplot2::element_text(size = 18),
      plot.title = element_text(hjust = 0.5, size = 22, face = "bold"),
      legend.title = element_blank(),
      legend.text = element_text(size = 18),
      legend.key.size = unit(1.2, "cm")
    ) +
    ggplot2::labs(
      x = "x coordinates",
      y = "y coordinates",
      title = paste0(gene, " - NUC-7738")
    ) +
    ggplot2::scale_fill_gradient(low = "white", high = "red", limits = c(0, 2), oob = scales::squish) +
    ggplot2::annotate("segment",
      x = 30000, xend = 30000 + scale_bar_whole,
      y = -35000, yend = -35000,
      color = "black", linewidth = 2, lineend = "round"
    ) +
    ggplot2::annotate("text",
      x = 30000 + scale_bar_whole / 2, y = -35500,
      label = "150 µm",
      color = "black", size = 8, hjust = 0.5, fontface = "bold"
    )
  
  whole_section_plots[[paste0(gene, "_DMSO")]] <- dmso_whole
  whole_section_plots[[paste0(gene, "_NUC7738")]] <- nuc_whole
}

# Combine whole section plots
whole_section_combined <- wrap_plots(whole_section_plots, ncol = 2)

# Save to supplementary
ggsave(
  filename = file.path(supplementary_figures, paste0("Supplementary_", gene_name, "_whole_section.png")),
  plot = whole_section_combined,
  width = 22, height = 10, dpi = 600, units = "in"
)

ggsave(
  filename = file.path(supplementary_figures, paste0("Supplementary_", gene_name, "_whole_section.pdf")),
  plot = whole_section_combined,
  width = 22, height = 10, device = cairo_pdf
)

print(whole_section_combined)
```

# 18. Session Info

```{r session-info}
sessionInfo()
```
